modules/gitrepo/config.go

```diff
@@ -0,0 +1,155 @@
+		AddDynamicArguments(key).
+		RunStdString(ctx, &git.RunOpts{Dir: repoPath(repo)})
+	if err != nil {
+		return "", err
+	}
+	if len(result) > 0 {
+		result = result[:len(result)-1] // remove trailing newline
+	}
+	return result, nil
+}
+
+func getRepoConfigLockKey(repoStoragePath string) string {
+	return "repo-config:" + repoStoragePath
+}
+
+// GitConfigAdd add a git configuration key to a specific value for the given repository.
+func GitConfigAdd(ctx context.Context, repo Repository, key, value string) error {
+	releaser, err := globallock.Lock(ctx, getRepoConfigLockKey(repo.RelativePath()))
+	if err != nil {
+		return err
+	}
+	defer releaser()
+
+	_, _, err = git.NewCommand("config", "--add").
+		AddDynamicArguments(key, value).
+		RunStdString(ctx, &git.RunOpts{Dir: repoPath(repo)})
+	return err
+}
+
+// GitConfigSet updates a git configuration key to a specific value for the given repository.
+// If the key does not exist, it will be created.
+// If the key exists, it will be updated to the new value.
+func GitConfigSet(ctx context.Context, repo Repository, key, value string) error {
+	releaser, err := globallock.Lock(ctx, getRepoConfigLockKey(repo.RelativePath()))
+	if err != nil {
+		return err
+	}
+	defer releaser()
+
+	_, _, err = git.NewCommand("config").
+		AddDynamicArguments(key, value).
+		RunStdString(ctx, &git.RunOpts{Dir: repoPath(repo)})
+	return err
+}
+
+type RemoteOption string
+
+const (
+	RemoteOptionMirrorPush  RemoteOption = "--mirror=push"
+	RemoteOptionMirrorFetch RemoteOption = "--mirror=fetch"
+)
+
+func GitRemoteAdd(ctx context.Context, repo Repository, remoteName, remoteURL string, options ...RemoteOption) error {
+	releaser, err := globallock.Lock(ctx, getRepoConfigLockKey(repo.RelativePath()))
+	if err != nil {
+		return err
+	}
+	defer releaser()
+
+	cmd := git.NewCommand("remote", "add")
+	if len(options) > 0 {
+		switch options[0] {
+		case RemoteOptionMirrorPush:
+			cmd.AddArguments("--mirror=push")
+		case RemoteOptionMirrorFetch:
+			cmd.AddArguments("--mirror=fetch")
+		default:
+			return errors.New("unknown remote option: " + string(options[0]))
+		}
+	}
+	_, _, err = cmd.
+		AddDynamicArguments(remoteName, remoteURL).
+		RunStdString(ctx, &git.RunOpts{Dir: repoPath(repo)})
+	return err
+}
+
+func GitRemoteRemove(ctx context.Context, repo Repository, remoteName string) error {
+	releaser, err := globallock.Lock(ctx, getRepoConfigLockKey(repo.RelativePath()))
+	if err != nil {
+		return err
+	}
+	defer releaser()
+
+	cmd := git.NewCommand("remote", "rm").AddDynamicArguments(remoteName)
+	_, _, err = cmd.RunStdString(ctx, &git.RunOpts{Dir: repoPath(repo)})
+	return err
+}
+
+// GitRemoteGetURL returns the url of a specific remote of the repository.
+func GitRemoteGetURL(ctx context.Context, repo Repository, remoteName string) (*giturl.GitURL, error) {
+	addr, err := git.GetRemoteAddress(ctx, repoPath(repo), remoteName)
+	if err != nil {
+		return nil, err
+	}
+	if addr == "" {
+		return nil, nil
+	}
```

Comment:
I believe this design is problematic. 

Developers have been used to that "if err == nil, then there is a non-nil return value"

If you do so, developers would forget to check "u == nil" frequently, and the nil check introduces unnecessary code.

Reply:
ae3c338be3